### ISSUE TITLE
Inconsistency in _assert_no_grad between nn.functional and nn

### ISSUE THREAD

**Issue by karandwivedi42** [pytorch/pytorch#7186]
In various losses, nn.functional does not have _assert_no_grad but Module does. I guess both have their advantages so maybe we can add a flag and make them uniform or document the fact that nn.functional can backpropogate through targets.

**Comment by zou3519** [comment]
A lot of those losses actually do return gradients for the targets now. I think the right fix here is to remove `_assert_no_grad `: autograd should warn if the gradients can't be computed for certain targets.
